### PR TITLE
fix(llm): fold non-leading system messages for models that reject them (fixes #3779)

### LINKED ISSUE #3779: OpenAI-compatible providers: fold non-leading system messages for chat templates that reject them (Qwen3.5 on vLLM returns 400 on every gptme request)
## Problem

gptme sends tool results and injected context as `system` messages in the middle of the conversation. Qwen3.5's stock chat template raises `System message must be at the beginning.` for any non-leading system message, so **every** request from gptme to a Qwen3.5 model served by vLLM (or any OpenAI-compatible server that applies that template) fails with HTTP 400. Observed 2026-09-09: 580/580 requests rejected during a gptme-eval run against `Qwen/Qwen3.5-0.8B` on vLLM 0.28.

Other model families with the same template rule will hit this too.

## Existing precedent

`gptme/llm/llm_openai.py` already has `_prep_o1`, which converts system messages (without `call_id`) to `user` messages wrapped in `<system>…</system>` for OpenAI reasoning models that don't support the system role.

## Proposal

- Keep the *first* system message as `system`.
- For OpenAI-compatible/local providers, fold every *non-leading* `system` message (no `call_id`) into a `user` message wrapped in `<system>…</system>`, the same way `_prep_o1` does, controlled by a model-meta flag (e.g. `supports_mid_system=False`) with an env override for custom endpoints (`GPTME_FOLD_SYSTEM_MESSAGES=1`).
- Tool results that are already `tool`-role (native format) are unaffected.

## Workaround used meanwhile

Serve vLLM with a copy of the Qwen3.5 template where the raise is replaced by rendering the message as a ChatML `system` turn. That matches how such data looks in gptme's own logs and in fine-tuning datasets derived from them (see the finetuning docs PR #3778).

Found while running the tool-format fine-tuning experiment; details in the gptme docs PR #3778.

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Problem

Qwen3.5's chat template raises `System message must be at the beginning.` for any system message that is not the first in the conversation. gptme injects tool results and context as `system` messages mid-conversation, causing every request to a Qwen3.5-on-vLLM endpoint to fail with HTTP 400.

Observed: 580/580 requests rejected during a gptme-eval run against `Qwen/Qwen3.5-0.8B` on vLLM 0.28. Other model families with the same template restriction are affected too.

## Fix

Adds `_fold_mid_system()` (modelled after the existing `_prep_o1`) and a `ModelMeta.supports_mid_system` flag (default `True` — backward-compatible):

- The **first** system message is kept as `system`.
- Every **subsequent** system message without a `call_id` is re-emitted as a `user` message wrapped in `<system>…</system>`, matching what `_prep_o1` does for OpenAI O1.
- Tool-result messages (system role + `call_id`) are **not** affected.

**Activation** (either works independently):
- `ModelMeta.supports_mid_system = False` for a known-restrictive model in the registry or a custom provider definition
- `GPTME_FOLD_SYSTEM_MESSAGES=1` env var for any custom/local endpoint without changing config

## Changes

- `gptme/llm/models/types.py`: add `supports_mid_system: bool = True` to `ModelMeta` and `_ModelDictMeta`
- `gptme/llm/llm_openai.py`: add `_fold_mid_system()`, wire into `_prepare_messages_for_api`
- `gptme/llm/models/listing.py`: serialize `supports_mid_system=False` in `model_to_dict`
- `tests/test_fold_mid_system.py`: 13 tests covering the folding logic, ModelMeta flag, env var activation, and integration with `_prepare_messages_for_api`

Fixes #3779

### CODE REVIEW

**Comment by TimeToBuildBob** [comment]
## Review guide (auto-posted)

Structured warm-up for reviewers — **what changed**, **what to run**, **where to look**.

### Key files
- `tests/test_fold_mid_system.py` (+251/-0, Δ251)
- `gptme/llm/llm_openai.py` (+42/-0, Δ42)
- `gptme/llm/models/listing.py` (+2/-0, Δ2)
- `gptme/llm/models/types.py` (+2/-0, Δ2)

### Suggested verification
- `uv run pytest tests/test_fold_mid_system.py -q`
- `make typecheck`

### Known risks / watch points
- No automatic high-risk tags; use file list + diff for judgment.

### Suggested review focus
- Confirm behavior matches the PR description acceptance criteria.
- Skim the largest diffs first (listed above).
- If CI is green, spot-check the highest-risk paths called out here.

_Generated by `scripts/github/pr-warmup-review-guide.py` for gptme/gptme#3780._

File: gptme/llm/models/types.py [hunk 1]
```diff
@@ -131,6 +131,7 @@ class ModelMeta:
         False  # models that can emit multiple tool calls in a single response
     )
     supports_strict_tools: bool = False  # models that support strict=True in tool schemas (OpenAI structured outputs)
+    supports_mid_system: bool = True  # whether the model/server accepts system messages that are not the first message in the conversation; set False for Qwen3.5 and similar chat templates that raise on non-leading system messages
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/llm/models/types.py:148]: Fixed in f81e4a939.

`get_model()` now infers `supports_mid_system=False` from Qwen3.5 model ids, so `local/Qwen/Qwen3.5-0.8B` activates folding without an env override or mocked metadata. Same inference is applied in OpenAI-compat listing. `GPTME_FOLD_SYSTEM_MESSAGES=1` remains for other restrictive templates.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
CI on `a42bd3624` failed: `model_to_dict` raised `AttributeError` on duck-typed models missing `supports_mid_system` (the `gptme-util models list --json` tests).

Pushed `9d116ec91` — `getattr(..., True)` so JSON listing doesn't crash. Local: the 3 previously-failing CLI tests plus `TestModelToDict` all pass.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
## 🤖 AI code review

Adds tests for non-leading system message folding and related ModelMeta flags, plus integration tests for message preparation.

### Safe to merge — no P0/P1 findings

Confidence 5/5

> ⚠️ **1 of 7 changed files were not fully read.** Claims that an identifier is never assigned, used, or called are discarded unless verification names only fully preloaded files. This coverage describes the bounded context preload; the diff was reviewed.

✅ **No findings.** The diff looks correct to me on this pass.

<details><summary><b>Files changed (2)</b> — the diff as I read it</summary>

- `tests/test_fold_mid_system.py` — Looks like a new or updated test file for system message folding.
- `tests/test_llm_models_resolution.py` — Contains tests for model resolution, including closest-match edge cases and provider routing.

<sub>Also changed, not described above: `gptme/llm/llm_openai.py`, `gptme/llm/models/__init__.py`, `gptme/llm/models/listing.py`, `gptme/llm/models/resolution.py`, `gptme/llm/models/types.py`</sub>

</details>

<sub>Reviewed `9d116ec917a8` · `openrouter/deepseek/deepseek-v4-flash-0731` · `llm` engine · 48s · [about this reviewer](https://github.com/gptme/gptme-contrib/blob/master/docs/ai-reviewer.md)</sub>

<details><summary><sub>Maintainer commands</sub></summary>

<sub>`@TimeToBuildBob review` (own line) — fresh review · `@TimeToBuildBob fix` — a worker acts on the findings. Once per comment; 👀 = received.</sub>

</details>

**Comment by TimeToBuildBob** [comment]
Self-merge not eligible here (policy gate: `gptme/llm/` provider-routing surface). CI green, Greptile 5/5, mergeable. Waiting for human review or merge.
